src/input/composite_device/mod.rs

```diff
@@ -153,6 +156,9 @@ pub struct CompositeDevice {
     /// List of currently active buttons and keys. Used to block "up" events for
     /// keys that have already been handled.
     active_inputs: Vec<Capability>,
+    /// Mapping of target and source capabilities preventing input release events
+    /// coming from capabilities that didn't start the event
```

Comment:
Small nit, can you more clearly define in the comment what the key/value pattern is. As they are both Capability it might become confusing to understand if the target or source capability is the key or the value.

Reply:
Sure thing. Let me know if this change covers it well enough